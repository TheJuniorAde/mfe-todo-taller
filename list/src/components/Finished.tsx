import React, { useContext, useMemo } from "react"
import { TodoContext } from "host/Context"
import { Item } from "./Item"
import { EmptyList } from "./EmptyList"

export const Finished: React.FC = () => {
  const context = useContext(TodoContext)
  const finishedItems = useMemo(
    () => context.list.filter((todo) => todo.status === "completed"),
    [context.list]
  )

  if (!context.list.length || !finishedItems.length) return <EmptyList />

  return (
    <ul role="list" className="divide-y divide-gray-100">
      {finishedItems.map((todo) => (
        <Item key={todo.id} todo={todo} updateTodo={context.toggleStatus} />
      ))}
    </ul>
  )
}
