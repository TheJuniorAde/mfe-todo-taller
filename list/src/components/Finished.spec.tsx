/**
 * @jest-environment jsdom
 */
import React, { createContext as mockContext } from "react"
import { render, fireEvent } from "@testing-library/react"
import { Finished } from "./Finished"
import { TodoContext, TodoItem } from "host/Context"

jest.mock("host/Context", () => ({ TodoContext: mockContext({}) }), {
  virtual: true,
})

const mockedItem: TodoItem = {
  description: "description",
  id: "id",
  finishedAt: new Date().toLocaleString("en-US"),
  createdAt: "",
  status: "completed",
}

describe("Finished", () => {
  it("renders correctly without tasks", () => {
    const { container } = render(
      <TodoContext.Provider
        value={{ list: [], updateData: jest.fn(), toggleStatus: jest.fn() }}
      >
        <Finished />
      </TodoContext.Provider>
    )

    expect(container.parentElement).toMatchSnapshot()
  })

  it("renders correctly without completed tasks", () => {
    const { container } = render(
      <TodoContext.Provider
        value={{
          list: [{ ...mockedItem, status: "incomplete" }],
          updateData: jest.fn(),
          toggleStatus: jest.fn(),
        }}
      >
        <Finished />
      </TodoContext.Provider>
    )

    expect(container.parentElement).toMatchSnapshot()
  })

  it("renders correctly with completed tasks", () => {
    const { container } = render(
      <TodoContext.Provider
        value={{
          list: [mockedItem],
          updateData: jest.fn(),
          toggleStatus: jest.fn(),
        }}
      >
        <Finished />
      </TodoContext.Provider>
    )

    expect(container.parentElement).toMatchSnapshot()
  })

  it("calls toggle functions", () => {
    const toggleMock = jest.fn()

    const { getByTestId } = render(
      <TodoContext.Provider
        value={{
          list: [mockedItem],
          updateData: jest.fn(),
          toggleStatus: toggleMock,
        }}
      >
        <Finished />
      </TodoContext.Provider>
    )

    fireEvent.click(getByTestId("list-item-label-id"))

    expect(toggleMock).toHaveBeenCalledWith(mockedItem)
  })
})
