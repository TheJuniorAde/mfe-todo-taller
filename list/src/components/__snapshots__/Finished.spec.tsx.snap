// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Finished renders correctly with completed tasks 1`] = `
<body>
  <div>
    <ul
      class="divide-y divide-gray-100"
      role="list"
    >
      <li
        class="flex justify-between gap-x-6 px-5 py-5 hover:bg-gray-100 hover:cursor-pointer"
        data-testid="list-item-id"
      >
        <label
          class="flex justify-between w-full hover:cursor-pointer"
          data-testid="list-item-label-id"
        >
          <div
            class="flex min-w-0 gap-x-4"
          >
            <input
              checked=""
              class="mr-2 leading-tight"
              type="checkbox"
            />
            <div
              class="min-w-0 flex-auto"
            >
              <p
                class="text-sm font-semibold leading-6 text-gray-900 line-through"
              >
                description
              </p>
              <p
                class="mt-1 truncate text-xs leading-5 text-gray-500"
              >
                Finished at 6/4/2024, 6:21:01 PM
              </p>
            </div>
          </div>
          <div
            class="hidden shrink-0 sm:flex sm:flex-col sm:items-end justify-center"
          >
            <div
              class="mt-1 flex items-center gap-x-1.5"
            >
              <div
                class="flex-none rounded-full bg-emerald-500/50 p-1"
              >
                <div
                  class="h-5 w-5 rounded-full bg-emerald-500/50"
                />
              </div>
              <p
                class="text-xs leading-5 text-gray-500"
              >
                completed
              </p>
            </div>
          </div>
        </label>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Finished renders correctly without completed tasks 1`] = `
<body>
  <div>
    <ul
      class="divide-y divide-gray-100"
      role="list"
    >
      <li
        class="flex justify-between gap-x-6 px-5 py-5 hover:bg-gray-100 "
      >
        <p
          class="block text-gray-700 text-sm font-bold"
        >
          No finished tasks
        </p>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Finished renders correctly without tasks 1`] = `
<body>
  <div>
    <ul
      class="divide-y divide-gray-100"
      role="list"
    >
      <li
        class="flex justify-between gap-x-6 px-5 py-5 hover:bg-gray-100 "
      >
        <p
          class="block text-gray-700 text-sm font-bold"
        >
          No finished tasks
        </p>
      </li>
    </ul>
  </div>
</body>
`;
